### PR TITLE
executor: Refine the implementation of `HashJoinExecutor`

### LINKED ISSUE #14466: Refine the implementation of `HashJoinExecutor`
## Feature Request

**Is your feature request related to a problem? Please describe:**

Now the implementation of TiDB's hash-join executor is rough and after profiling, we found there are some issues can be resolved to get better performance:
image
- [ ] the method `joinMatchedProbeSideRow2Chunk` uses the method `NewIterator4Slice` to create a lot of iterators whereas these iterators can be reused.
- [ ] some slices created in `GetMatchRowsAndPtrs` can be reused.
- [ ] we can try to change the return type of `entryStore.get` from structs to pointers which may reduce memory copy(but it may cause some GC problems so we need some tests after update).

If you have any other ideas, welcome to discuss it with us in our [SIG-Exec channel](https://tidbcommunity.slack.com/archives/CMRD79DRR) or in this issue by using comments.

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?
Close #14466 
Refine the implementation of `HashJoinExecutor` to get better performance. 

### What is changed and how it works?
1. Reuse the iterators in the method `joinMatchedProbeSideRow2Chunk`.
2. Reuse some slices in the method `GetMatchRowsAndPtrs`.
3. ~~Change the return type of `entryStore.get` from structs to pointers which may reduce memory copy.~~ After benchmarking, there is no performance improvement here.

And here is the benchmark:
```
name                                                                                                        old time/op    new time/op    delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-4     1.55s ± 8%     1.43s ± 3%   -7.62%  (p=0.001 n=8+9)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4       292ms ± 7%     279ms ± 4%     ~     (p=0.105 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-4        1.13s ±23%     1.10s ±12%     ~     (p=0.684 n=10+10)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4                  99.2µs ±14%   104.8µs ±13%     ~     (p=0.105 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-4           49.3ms ± 9%    43.5ms ± 3%  -11.69%  (p=0.000 n=9+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4             43.8ms ± 8%    37.6ms ± 1%  -14.16%  (p=0.000 n=10+9)

name                                                                                                        old alloc/op   new alloc/op   delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-4    2.46GB ± 0%    2.45GB ± 0%   -0.23%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4       305MB ± 0%     300MB ± 0%   -1.83%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-4        890MB ± 0%     885MB ± 0%   -0.63%  (p=0.000 n=10+10)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4                  49.5kB ± 0%    49.6kB ± 0%   +0.13%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-4           18.1MB ± 0%    12.5MB ± 0%  -31.02%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4             16.5MB ± 0%    10.9MB ± 0%  -34.00%  (p=0.000 n=10+10)

name                                                                                                        old allocs/op  new allocs/op  delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-4      812k ± 0%      512k ± 0%  -36.93%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4        412k ± 0%      112k ± 0%  -72.78%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-4        1.61M ± 0%     1.31M ± 0%  -18.60%  (p=0.000 n=10+10)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4                     253 ± 0%       243 ± 0%   -3.95%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-4             612k ± 0%      312k ± 0%  -49.01%  (p=0.000 n=10+9)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-4               412k ± 0%      112k ± 0%  -72.82%  (p=0.000 n=10+10)
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
- executor: Refine the implementation of `HashJoinExecutor`

### CODE REVIEW

**Comment by Reminiscent** [comment]
@qw4990 @wshwsh12 @SunRunAway  PTAL

File: executor/hash_table.go [hunk 1]
```diff
@@ -123,14 +123,21 @@ func newHashRowContainer(sCtx sessionctx.Context, estCount int, hCtx *hashContex
...
+	if lenInnerPtrs == 0 {
 		return
 	}
-	matched = make([]chunk.Row, 0, len(innerPtrs))
+	if matched == nil || cap(*matched) < lenInnerPtrs {
+		*matched = make([]chunk.Row, 0, lenInnerPtrs)
+	}
+	if matchedPtrs == nil || cap(*matchedPtrs) < lenInnerPtrs {
+		*matchedPtrs = make([]chunk.RowPtr, 0, lenInnerPtrs)
+	}
```
Comment (wshwsh12) [inline executor/hash_table.go:139]: We can merge the two check.

**Comment by jianzhiyao** [comment]
Why close this pr？@reminiscent

**Pull request closed without merging**